web/source/osk/oskManager.ts

```diff
@@ -191,69 +189,54 @@ namespace com.keyman.osk {
-        if(activeKeyboard && activeKeyboard.layout(device.formFactor as utils.FormFactor)) {
-          this._GenerateVisualKeyboard(activeKeyboard);
-        } else if(!activeKeyboard) {
-          this._GenerateVisualKeyboard(null);
-        } else { //The following code applies only to preformatted 'help' such as SIL EuroLatin
-          //osk.ddOSK = false;
-          const layout = this.desktopLayout = new layouts.TargetedFloatLayout();
-          layout.attachToView(this);
-          this._Box.appendChild(layout.titleBar.element);
-          this._Box.appendChild(this.banner.element);
-
-          //Add content
-          var Ldiv = util._CreateElement('div');
-          Ldiv.className='kmw-osk-static';
-          Ldiv.innerHTML = Lhelp;
-          this._Box.appendChild(Ldiv);
-          if(activeKeyboard.hasHelpHTML) {
-            activeKeyboard.insertHelpHTML(this._Box);
-          }
-        }
+      let kbdView: KeyboardView = this._GenerateKeyboardView(activeKeyboard);
+      this._Box.appendChild(kbdView.element);
+      if(kbdView instanceof VisualKeyboard) {
+        this.vkbd = kbdView;
+      }
+      kbdView.postInsert();
 
-        if(this.desktopLayout) {
-          this.desktopLayout.titleBar.setTitleFromKeyboard(activeKeyboard);
+      // Add footer element to OSK only for desktop browsers
+      if(this.desktopLayout) {
+        if(kbdView instanceof VisualKeyboard) {
+          this._Box.appendChild(layout.resizeBar.element);
         }
+        // For other devices, adjust the object heights, allowing for viewport scaling
+      } else {
+        this.vkbd.adjustHeights(this.getKeyboardHeight());
+
+        var b: HTMLElement = this._Box, bs=b.style;
+        bs.height=bs.maxHeight=this.vkbd.computedAdjustedOskHeight(this.getHeight())+'px';
       }
 
+      // END:  construction of the actual internal layout for the overall OSK
+
       // Correct the classname for the (inner) OSK frame (Build 360)
       var kbdID: string = (activeKeyboard ? activeKeyboard.id.replace('Keyboard_','') : '');
       if(keymanweb.isEmbedded && kbdID.indexOf('::') != -1) {
         // De-namespaces the ID for use with CSS classes.
         // Assumes that keyboard IDs may not contain the ':' symbol.
         kbdID = kbdID.substring(kbdID.indexOf('::') + 2);
       }
-      var innerFrame=<HTMLDivElement> this._Box.firstChild,
-        kbdClass = ' kmw-keyboard-' + kbdID;
-      if(innerFrame.id == 'keymanweb_title_bar') {
-        // Desktop order is title_bar, banner_container, inner-frame
-        innerFrame=<HTMLDivElement> innerFrame.nextSibling.nextSibling;
-      } else if (innerFrame.id == 'keymanweb_banner_container') {
-        innerFrame=<HTMLDivElement> innerFrame.nextSibling;
-      }
-      innerFrame.className = 'kmw-osk-inner-frame' + kbdClass;
+
+      var kbdClassSuffix = ' kmw-keyboard-' + kbdID;
+      kbdView.element.className = kbdView.element.className + kbdClassSuffix;
```

Comment:
`kbdView.element` is equivalent to the old `innerFrame` - but this time, we don't need to search `_Box`'s hierarchy to find it.